File: aizu/Introduction-to-Algorithms-and-Data-Structures/ALDS1_3_B_Queue/ALDS1_3_B.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

type Process struct {
	Name string
	Time int
}

type Queue struct {
	xs   [100001]*Process
	head int
	tail int
}

func (q *Queue) enqueue(x *Process) {
	q.xs[q.tail] = x
	if q.tail >= len(q.xs)-1 {
		q.tail = 0
	} else {
		q.tail++
	}
}

func (q *Queue) dequeue() *Process {
	x := q.xs[q.head]
	if q.head >= len(q.xs)-1 {
		q.head = 0
	} else {
		q.head++
	}
	return x
}

func nextInt(sc *bufio.Scanner) int {
	sc.Scan()
	n, err := strconv.Atoi(sc.Text())
	if err != nil {
		panic(err)
	}
	return n
}

func nextStr(sc *bufio.Scanner) string {
	sc.Scan()
	s := sc.Text()
	return s
}

func main() {
	sc := bufio.NewScanner(os.Stdin)
	sc.Split(bufio.ScanWords)
	n := nextInt(sc)
	q := nextInt(sc)

	var queue Queue
	for i := 0; i < n; i++ {
		queue.enqueue(&Process{
			Name: nextStr(sc),
			Time: nextInt(sc),
		})
	}

	sum := 0
	for n > 0 {
		p := queue.dequeue()
		if p.Time > q {
			sum = sum + q
			p.Time = p.Time - q
			queue.enqueue(p)
		} else {
			sum = sum + p.Time
			n--
			fmt.Print(p.Name)
			fmt.Print(" ")
			fmt.Println(sum)
		}
	}
}

File: aizu/Introduction-to-Algorithms-and-Data-Structures/ALDS1_3_B_Queue/ALDS1_3_B.rb
class Queue
  def initialize
    @arr = []
    @head = 0
    @tail = 0
  end

  def enque(x)
    @arr[@tail] = x
    @tail = @tail + 1
  end

  def deque
    x = @arr[@head]
    @head = @head + 1
    x
  end

  def size
    @tail - @head
  end
end

def roundrobin(q, queue)
  finished = []
  t = 0
  while queue.size != 0
    process = queue.deque
    if (process[1].to_i <= q)
      t = t + process[1].to_i
      process[1] = 0
      finished.push("#{process[0]} #{t}")
    else
      t = t + q
      process[1] = process[1].to_i - q
      queue.enque(process)
    end
  end

  puts finished.join("\n")
end



as = STDIN.gets.split(' ')

n = as[0].to_i
q = as[1].to_i

queue = Queue.new
n.times.each {
  queue.enque(STDIN.gets.split(' '))
}

roundrobin(q, queue)

